[clangd] Carefully handle PseudoObjectExprs for inlay hints

zyn0217 opened the pull request:
This closes https://github.com/clangd/clangd/issues/1813.

Not all subexpressions for a PseudoObjectExpr are interesting, and they probably mess up inlay hints.

zyn0217 committed:
[clangd] Don't show inlay hints for PseudoObjectExprs in C++

This closes https://github.com/clangd/clangd/issues/1813.

PseudoObjectExprs in C++ are currently not very interesting but probably
mess up inlay hints.

zyn0217 committed:
Don't make the patch C++-specific

zyn0217 commented:
On second thought I think I should not kill all the expressions involving `PseudoObjectExpr`: I'm not sure if our ObjC / CUDA support does need that for presenting useful hints.

zyn0217 committed:
Only apply the restriction to __builtin_dump_struct

HighCommander4 reviewed (commented):
Is it not possible to tell based on the `CallExpr` only (rather than checking an ancestor of it during traversal of the tree) that "this `CallExpr` is not explicitly written in the source"?

zyn0217 commented:
> Is it not possible to tell based on the `CallExpr` only (rather than checking an ancestor of it during traversal of the tree) that "this `CallExpr` is not explicitly written in the source"?

I'm not quite sure - Looking into the construction process for a `__builtin_dump_struct` expression,
https://github.com/llvm/llvm-project/blob/4d5a8ccf25a6c43930c332e75ecead287778af6b/clang/lib/Sema/SemaChecking.cpp#L482-L509

the part that makes the call unordinary might be the nullish scope. I don't know if that implies "the expression is not written by users"; even so, extracting the scope of a call expression then performing the check doesn't look much more straightforward than this way.

HighCommander4 commented:
I see; it's a bit unfortunate that such an "invented" call expression gets an ordinary `SourceLocation`. I would have hoped it gets a source location in some sort of virtual buffer like the [rewritten form of `>>`](https://github.com/clangd/clangd/issues/871#issuecomment-933178468), or macro expansions.

---

What do you think about the following idea: override the traversal of `PseudoObjectExpr`, but rather than traversing [both the `getSyntacticForm()` and the semantic expressions](https://searchfox.org/llvm/rev/40671bbdefb6ff83e2685576a3cb041b62f25bbe/clang/include/clang/AST/RecursiveASTVisitor.h#2573-2584), only traverse the `getSyntacticForm()`?

That leave the door open to a future enhancement where the arguments in the syntactic form (the `__builtin_dump_struct`) could get parameter hints, which might be useful for a reader to understand what the builtin does.

HighCommander4 commented:
> I see; it's a bit unfortunate that such an "invented" call expression gets an ordinary `SourceLocation`.

Out of curiosity, I dialled into today's LLVM [office hours](https://llvm.org/docs/GettingInvolved.html#office-hours) and asked Aaron Ballman about this. He also did not see an obvious way to tell apart the invented `CallExpr`s from regular `CallExpr`s, but suggested that it could be reasonable to just change the code for building the invented `CallExpr`s to give them an invalid SourceLocation (and then our inlay hints code should filter them out without further changes).

zyn0217 commented:
> Out of curiosity, I dialled into today's LLVM office hours and asked Aaron Ballman about this

Thank you so much for making the effort on this issue. :)

> ...just change the code for building the invented CallExprs to give them an invalid SourceLocation

Interesting idea and I'll explore it.

zyn0217 committed:
Override TraversePseudoObjectExpr

zyn0217 commented:
Since the other PR turned out to be infeasible, I'm overriding `TraversePseudoObjectExpr` now.

HighCommander4 commented on clang-tools-extra/clangd/unittests/InlayHintTests.cpp:
```diff
@@ -1724,6 +1724,33 @@ TEST(InlayHints, RestrictRange) {
+    struct S {
+      __declspec(property(get=GetX, put=PutX)) int x[];
+      int GetX(int y, int z) { return 42 + y; }
+      void PutX(int y) { x = $one[[y]]; } // FIXME: Undesired `x = y: y` for this ill-formed expression.
```
This is a bit confusing, since the fact that the expression is ill-formed is not related to the FIXME.

I would suggest moving the `x = y;` to some other function (where it's not ill-formed).

(Also mention in a comment that this builds a PseudoObjectExpr here too.)

zyn0217 replied:
Sorry, I'm a bit inaccurate here: The assignment expression in `PutX` is not ill-formed but would expand to a recursive call to `PutX` since we're assigning to the member with the very setter. (Of course, this would end up crashing if we [run the program](https://gcc.godbolt.org/z/e3fW3W73a).)

I think I was suggesting that this is a bad case for supporting such an extension: the assignment expression could be of a syntactic form with an inappropriate hint location though.

zyn0217 replied:
It's nice to wipe out these extra hints if it won't take too many steps. The AST shows,
```
`-PseudoObjectExpr 'void'
  |- (Syntactic) BinaryOperator
  `- (Semantic) CXXMemberCallExpr 'void'
```
I wonder if it is possible to enumerate all the bad cases we don't want (e.g. Don't walk down the expression if we encounter a `BinaryOperator` in `PseudoObjectExpr`.)

HighCommander4 replied:
I'm not sure what other bad cases there might be, but excluding cases where the syntactic form is a `BinaryOperator` to avoid this case sounds fine to me.

HighCommander4 commented on clang-tools-extra/clangd/unittests/InlayHintTests.cpp:
```diff
@@ -1724,6 +1724,33 @@ TEST(InlayHints, RestrictRange) {
+      S s;
+      __builtin_dump_struct(&s, printf); // Not `Format: __builtin_dump_struct()`
+      printf($Param[["Hello, %d"]], 42); // Normal calls are not affected.
+      return s.x[ $two[[1]] ][ $three[[2]] ]; // `x[y: 1][z: 2]`
```
Thanks, this is an interesting use case for not just skipping the semantic forms of all PseudoObjectExprs (which was my initial thought)

Since it's not obvious that this builds a PseudoObjectExpr, can you please add a comment:

```
// This builds a PseudoObjectExpr, but here it's useful for show the
// arguments from the semantic form.
```

HighCommander4 commented on clang-tools-extra/clangd/InlayHints.cpp:
```diff
@@ -589,6 +589,24 @@ class InlayHintVisitor : public RecursiveASTVisitor<InlayHintVisitor> {
+      // possibly get parameter names.
+      return RecursiveASTVisitor<InlayHintVisitor>::TraverseStmt(
+          E->getSyntacticForm());
+    // FIXME: Shall we ignore semantic forms for other pseudo object
```
Maybe all PseudoObjectExprs where the syntactic form is a CallExpr with a builtin callee would be reasonable?

zyn0217 replied:
I think this depends on our attitude towards parameter hints for other pseudo-expression usages. Take the getter case from the test for example, the expression for `S().x[1][2]` turns out to be
```
`-PseudoObjectExpr 'int'
 |- (Syntatic) MSPropertySubscriptExpr '<pseudo-object type>' lvalue
 `- (Semantic) CXXMemberCallExpr 'int'
```
There's no chance to see hints for the arguments if we neglect all the syntactic forms being `CallExpr`s. Besides, there are still some similar usages in ObjC / OpenMP that I'm not familiar with. 
Anyway, if we agree to sacrifice the chances to see these hints, I'd love to move forward with all `CallExpr`s within a pseudo-expression being dropped. Otherwise, I think it's better to leave it to the next bug report / feature request. What do you think?

HighCommander4 replied:
This is not a case where the **syntactic** form is a `CallExpr` (here the syntactic form is an `MSPropertySubscriptExpr`), so here the hints would be retained.

zyn0217 replied:
Oops. sorry for messing up the types. :-(

zyn0217 committed:
Address some comments on the test.

zyn0217 committed:
Append a comment

HighCommander4 commented on clang-tools-extra/clangd/unittests/InlayHintTests.cpp:
```diff
@@ -1724,6 +1724,36 @@ TEST(InlayHints, RestrictRange) {
+      __declspec(property(get=GetX, put=PutX)) int x[];
+      int GetX(int y, int z) { return 42 + y; }
+      // FIXME: Undesired hint `x = y: y`. This builds a PseudoObjectExpr too.
+      void PutX(int y) { x = $one[[y]]; }
```
While it may not be ill-formed, I think the fact that this causes an infinite recursion at runtime is a distraction for someone reading the test.

I would suggest instead something like:

```
void PutX(int y) {}

// FIXME: Undesired hint `x = y: y`. This builds a PseudoObjectExpr too.
void SomeOtherFunction(int y) { x = $one[[y]]; }
```

zyn0217 committed:
Address review comments

zyn0217 commented:
Sorry for putting this off; I've updated it and please take another look.
I also changed the PR title since I feel this is not particular to `__builtin_dump_struct`.

HighCommander4 reviewed (approved):
LGTM, thank you!
